x-pack/metricbeat/module/gcp/metrics/redis/metadata.go

```diff
@@ -0,0 +1,211 @@
+
+	if metadata.machineType != "" {
+		lastIndex := strings.LastIndex(metadata.machineType, "/")
+		_, _ = metadataCollectorData.ECS.Put(gcp.ECSCloudMachineTypeKey, metadata.machineType[lastIndex+1:])
+	}
+
+	metadata.Metrics = metadataCollectorData.Labels[gcp.LabelMetrics]
+	metadata.System = metadataCollectorData.Labels[gcp.LabelSystem]
+
+	if metadata.User != nil {
+		metadataCollectorData.Labels[gcp.LabelUser] = metadata.User
+	}
+
+	return metadataCollectorData, nil
+}
+
+// instanceMetadata returns the labels of an instance
+func (s *metadataCollector) instanceMetadata(ctx context.Context, instanceID, region string) (*redisMetadata, error) {
+	instance, err := s.instance(ctx, instanceID)
+	if err != nil {
+		return nil, fmt.Errorf("error trying to get data from instance '%s' %w", instanceID, err)
+	}
+
+	metadata := &redisMetadata{
+		instanceID: instanceID,
+		region:     region,
+	}
+
+	if instance == nil {
+		s.logger.Debugf("couldn't get instance '%s' call ListInstances API", instanceID)
+		return metadata, nil
+	}
+
+	if instance.Name != "" {
+		parts := strings.Split(instance.Name, "/")
+		metadata.instanceName = parts[len(parts)-1]
+	}
+
+	if instance.Labels != nil {
+		metadata.User = instance.Labels
+	}
+
+	if instance.Tier.String() != "" {
+		metadata.machineType = instance.Tier.String()
+	}
+
+	return metadata, nil
+}
+
+// instance returns data from an instance ID using the cache or making a request
+func (s *metadataCollector) instance(ctx context.Context, instanceID string) (*redispb.Instance, error) {
+	s.getInstances(ctx)
+
+	instance, ok := s.instances[instanceID]
+	if ok {
+		return instance, nil
+	}
+
+	s.instances = make(map[string]*redispb.Instance)
+
+	return nil, nil
+}
+
+func (s *metadataCollector) instanceID(ts *monitoringpb.TimeSeries) string {
+	if ts.Resource != nil && ts.Resource.Labels != nil {
+		return ts.Resource.Labels[gcp.TimeSeriesResponsePathForECSInstanceID]
+	}
+
+	return ""
+}
+
+func (s *metadataCollector) instanceRegion(ts *monitoringpb.TimeSeries) string {
+	if ts.Resource != nil && ts.Resource.Labels != nil {
+		return ts.Resource.Labels["region"]
+	}
+
+	return ""
+}
+
+func (s *metadataCollector) getInstances(ctx context.Context) {
+	if len(s.instances) > 0 {
+		return
+	}
+
+	s.logger.Debug("get redis instances with ListInstances API")
+
+	client, err := redis.NewCloudRedisClient(ctx, s.opt...)
+	if err != nil {
+		s.logger.Errorf("error getting client from redis service: %v", err)
+		return
+	}
+
+	defer client.Close()
+
+	// Use locations - (wildcard) to fetch all instances.
+	// https://pkg.go.dev/cloud.google.com/go/redis@v1.10.0/apiv1#CloudRedisClient.ListInstances
+	it := client.ListInstances(ctx, &redispb.ListInstancesRequest{
+		Parent: fmt.Sprintf("projects/%s/locations/-", s.projectID),
+	})
+	for {
+		instance, err := it.Next()
+		if errors.Is(err, iterator.Done) {
+			break
+		}
+
+		if err != nil {
+			s.logger.Errorf("redis ListInstances error: %v", err)
+			continue
```

Comment:
Just to double check, are you sure it's `continue` and not `break` here? How would the loop stop when an error occurs?